azure/scope/workload_identity.go

```diff
@@ -0,0 +1,157 @@
+|AZURE_FEDERATED_TOKEN_FILE | The path of the projected service account token file. |
+|-----------------------------------------------------------------------------------|
+
+In addition to the service account, it also projects a signed service account token to the
+workload's volume(in this case the capz pod). The volume name is `azure-identity-token`
+which is mounted at path `/var/run/secrets/azure/tokens/azure-identity-token` to the pod.
+
+*/
+
+const (
+	// AzureFedratedTokenFileEnvKey is the env key for AZURE_FEDERATED_TOKEN_FILE.
+	AzureFedratedTokenFileEnvKey = "AZURE_FEDERATED_TOKEN_FILE"
+	// AzureClientIDEnvKey is the env key for AZURE_CLIENT_ID.
+	AzureClientIDEnvKey = "AZURE_CLIENT_ID"
+	// AzureTenantIDEnvKey is the env key for AZURE_TENANT_ID.
+	AzureTenantIDEnvKey = "AZURE_TENANT_ID"
+	// AzureTokenFilePath is the path of the projected token.
+	AzureTokenFilePath = "/var/run/secrets/azure/tokens/azure-identity-token" // #nosec G101
+)
+
+type workloadIdentityCredential struct {
+	assertion string
+	file      string
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+// WorkloadIdentityCredentialOptions contains the configurable options for azwi.
+type WorkloadIdentityCredentialOptions struct {
+	azcore.ClientOptions
+	ClientID      string
+	TenantID      string
+	TokenFilePath string
+}
+
+// NewWorkloadIdentityCredentialOptions returns an empty instance of WorkloadIdentityCredentialOptions.
+func NewWorkloadIdentityCredentialOptions() *WorkloadIdentityCredentialOptions {
+	return &WorkloadIdentityCredentialOptions{}
+}
+
+// WithClientID sets client ID to WorkloadIdentityCredentialOptions.
+func (w *WorkloadIdentityCredentialOptions) WithClientID(clientID string) *WorkloadIdentityCredentialOptions {
+	w.ClientID = clientID
+	return w
+}
+
+// WithTenantID sets tenant ID to WorkloadIdentityCredentialOptions.
+func (w *WorkloadIdentityCredentialOptions) WithTenantID(tenantID string) *WorkloadIdentityCredentialOptions {
+	w.TenantID = tenantID
+	return w
+}
+
+// GetProjectedTokenPath return projected token file path from the env variable.
+func GetProjectedTokenPath() (string, error) {
+	tokenPath := os.Getenv(AzureFedratedTokenFileEnvKey)
+	if strings.TrimSpace(tokenPath) == "" {
+		return AzureTokenFilePath, nil
+	}
+	return tokenPath, nil
+}
+
+// WithDefaults sets token file path. It also sets the client tenant ID from injected env in
+// case empty values are passed.
+func (w *WorkloadIdentityCredentialOptions) WithDefaults() (*WorkloadIdentityCredentialOptions, error) {
+	tokenFilePath, err := GetProjectedTokenPath()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get token file path for identity")
+	}
+	w.TokenFilePath = tokenFilePath
+
+	// Fallback to using client ID from env variable if not set.
+	if strings.TrimSpace(w.ClientID) == "" {
+		w.ClientID = os.Getenv(AzureClientIDEnvKey)
+		if strings.TrimSpace(w.ClientID) == "" {
```

Comment:
Maybe store the value after `strings.TrimSpace` for consistency?

Reply:
I will have to anyways again read w.ClientID, hence like that. 
Or I did not get what you are suggesting.

Reply:
do `string.TrimSpace` before storing it in `w.ClientID`. That way it's consistently trimmed before storing and you can avoid doing `strings.TrimSpace` every time you access it. This applies to all the other values that have `strings.TrimSpace` in this file.